### ISSUE TITLE
Add basic TimeZone support to Dates stdlib

### ISSUE THREAD

**Issue by simonbyrne** [JuliaLang/julia#51341]
Currently `DateTime` is completely time-zone unaware: this simplicity was intentional (as described by @quinnj in https://github.com/JuliaLang/julia/pull/49700#issuecomment-1547833304), with additional functionality pushed to [TimeZones.jl](https://github.com/JuliaTime/TimeZones.jl).

The problem is that many packages skip time zone support altogether, or implement it in inconsistent ways: for example, `now()` returns a local `DateTime` by default, whereas TOML.jl [treats all `DateTime`s as UTC](https://github.com/JuliaLang/julia/blob/a61d1b47a68297704814188a3509c011ec2a8fa1/stdlib/TOML/src/print.jl#L92). 

Even packages which aren't in the stdlib are reluctant to use TimeZones.jl, resulting in bugs or less-than-useful information:
- Scratch.jl [prints local times as UTC](https://github.com/JuliaPackaging/Scratch.jl/issues/28)
- PkgServer.jl [logs output in the local time](https://github.com/JuliaPackaging/PkgServer.jl/blob/5d314f7bb76f27557a242710725cc4a4bbbcd688/bin/run_server.jl#L37)

What I would like to propose is to add basic UTC and UTC + offset support into stdlib (basically functionality which doesn't require looking up a timezone database). This would be sufficient to fully support the [TOML datetime](https://toml.io/en/v1.0.0#offset-date-time) and many other applications.

 I was thinking something like the following
```julia
struct DateTimeZoned{TZ<:TimeZone} <: AbstractDateTime
    instant::UTInstant{Millisecond}
    zone::TZ
end
```

We already have `UTC <: TimeZone` defined, so we could also define
```julia
struct UTCOffset <: TimeZone
    offset::Millisecond # or could use Minute?
end
```

Similarly we could expand the `datetime""` macro to support the `Z` (which would return a `DateTimeZoned{UTC}` and `+00:00` suffixes (which would return a `DateTimeZoned{UTCOffset}`.

Like our current `DateTime`, we would continue to ignore leap seconds (as basically every other piece of software does).

This would fix the following issues:
- https://github.com/JuliaLang/julia/issues/23049
- https://github.com/JuliaLang/julia/issues/27811
- https://github.com/JuliaLang/julia/issues/49821

**Comment by quinnj** [comment]
I'd be in favor of this. cc: @omus

**Comment by simonbyrne** [comment]
Also cc @omus

**Comment by simonbyrne** [comment]
(I would also be in favor of renaming `UTInstant` to `POSIXInstant` or something like that, since what we're doing doesn't really have any correspondence to UT in any way, but this is less of an issue)

**Comment by giordano** [comment]
> ```julia
> struct UTCOffset <: TimeZone
>     offset::Millisecond # or could use Minute?
> end
> ```

There have been timezones with non-integer offsets in minutes in the past, for example: https://en.wikipedia.org/wiki/Moscow_Time#History. Milliseconds should be a safer choice

**Comment by tecosaur** [comment]
I wonder if there's some value in mentioning https://github.com/BurntSushi/jiff, which seems to take an approach rather in line with (what I'd hope at least) Julia stdlibs aim for:

> The most important design goal of Jiff is to be a high level datetime library that makes it hard to do the wrong thing. Second to that is performance.

There are a number of interesting choices in the library, such of the use of nanoseconds not microseconds to represent instances, only bundling the timezone DB on windows and using `/usr/share/zoneinfo` / `$TZDIR` on Linux, and supporting eras (BCE, CE).

**Comment by omus** [comment]
I'm not opposed to having this but it would be nice to avoid having a `DateTimeZoned` defined in Dates.jl and `ZonedDateTime` in TimeZones.jl. I think having both would lead to confusion.

> Even packages which aren't in the stdlib are reluctant to use TimeZones.jl,

Do we know why these packages are reluctant to depend on TimeZones.jl? Since the creation of TZJData.jl we've switched to artifacts for tzdata information which eliminated processing done previously during package initialization. If there are other concerns we can possibly address those directly TimeZones.jl

> There have been timezones with non-integer offsets in minutes in the past, for example: https://en.wikipedia.org/wiki/Moscow_Time#History. Milliseconds should be a safer choice

The initial offsets for most time zones in the IANA tz database include a offset with seconds. I've not seen anything with a higher resolution than seconds though.

**Comment by simonbyrne** [comment]
@omus Honestly I think it's because TimeZones.jl is overkill: they don't need actual time zones, just UTC time.

Thinking this through a bit more, I think simply doing something like
```
struct DateTimeUTC <: AbstractDateTime
    instant::UTInstant{Millisecond}  # or UTCInstant?
    offset::Millisecond
end
```
might be sufficient.

**Comment by omus** [comment]
I'd be more in favor of having the more limited scope of `DateTimeUTC`. Do you even need the offset in your proposed type? You could just do:

```julia
struct DateTimeUTC <: AbstractDateTime
    instant::UTInstant{Millisecond}  # or UTCInstant?
end
```

Effectively we would just be defining a type to be explicit that this datetime is in UTC. We would probably want to disallow conversion between `DateTimeUTC` and `DateTime` to avoid confusion when working with these types.

> Honestly I think it's because TimeZones.jl is overkill: they don't need actual time zones, just UTC time.

Another solution we could consider is splitting out `ZonedDateTime` from TimeZones.jl and having it exist in a separate package which doesn't depend on TimeZones. This would support fixed UTC offsets and could perform basic parsing of string with the `Z` suffix. I suspect that the packages which are avoiding TimeZones.jl are just trying to avoid dependencies outside of the stdlibs though.

**Comment by simonbyrne** [comment]
> I'd be more in favor of having the more limited scope of `DateTimeUTC`. Do you even need the offset in your proposed type? You could just do:

Personally I think it's worth having offset support: my thinking was to support [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601#Time_offsets_from_UTC)

**Comment by DilumAluthge** [comment]
> I'd be more in favor of having the more limited scope of `DateTimeUTC`. Do you even need the offset in your proposed type? You could just do:
> 
> struct DateTimeUTC <: AbstractDateTime
>     instant::UTInstant{Millisecond}  # or UTCInstant?
> end
> Effectively we would just be defining a type to be explicit that this datetime is in UTC. We would probably want to disallow conversion between `DateTimeUTC` and `DateTime` to avoid confusion when working with these types.
> 
> > Honestly I think it's because TimeZones.jl is overkill: they don't need actual time zones, just UTC time.
> 
> Another solution we could consider is splitting out `ZonedDateTime` from TimeZones.jl and having it exist in a separate package which doesn't depend on TimeZones. This would support fixed UTC offsets and could perform basic parsing of string with the `Z` suffix. I suspect that the packages which are avoiding TimeZones.jl are just trying to avoid dependencies outside of the stdlibs though.

This solution seems like it would cover a lot of use cases.

I'm always a little wary of adding more content to Base or a stdlib, so the idea of a separate slimmed-down package (with no build step or artifacts) seems like a good approach.

**Comment by simonbyrne** [comment]
I disagree: we still wouldn't be able to [parse TOML correctly](https://toml.io/en/v1.0.0#offset-date-time), and there are enough cases where local times are used (see linked issues), that people will just go back to dropping the offsets.
